include/h2o/tunnel.h

```diff
@@ -27,9 +27,53 @@ extern "C" {
 #endif
 
 typedef struct st_h2o_tunnel_t h2o_tunnel_t;
+typedef struct st_h2o_tunnel_endpoint_t h2o_tunnel_endpoint_t;
+typedef void (*h2o_tunnel_endpoint_open_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_endpoint_t *end);
+typedef void (*h2o_tunnel_endpoint_send_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_endpoint_t *end, h2o_iovec_t *bufs, size_t bufcnt,
+                                        int is_final);
+typedef void (*h2o_tunnel_endpoint_on_peer_send_complete_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_endpoint_t *end, h2o_tunnel_endpoint_t *peer);
+typedef void (*h2o_tunnel_endpoint_close_cb)(h2o_tunnel_t *tunnel, h2o_tunnel_endpoint_t *end, const char *err);
 
-h2o_tunnel_t *h2o_tunnel_establish(h2o_context_t *ctx, h2o_socket_t *sock1, h2o_socket_t *sock2, h2o_timeout_t *timeout);
-void h2o_tunnel_break(h2o_tunnel_t *tunnel);
+typedef struct st_h2o_tunnel_endpoint_callbacks_t {
+    /**
+     * called when tunnel is initialized
+     */
+    h2o_tunnel_endpoint_open_cb open;
+    /**
+     * called when the peer wants to send data to this end
+     */
+    h2o_tunnel_endpoint_send_cb send;
+    /**
+     * called when the peer completed to send the data (maybe NULL)
+     */
+    h2o_tunnel_endpoint_on_peer_send_complete_cb on_peer_send_complete;
+    /**
+     * called when tunnel gets broken
```

Comment:
Does "broken" mean that it was reset? I would clarify how the connection gets closed (both normally and abruptly) as well as the relationship with the `is_final` flag of the send callback.